### PR TITLE
Feat(template): Added scikit as an option for build system

### LINKED ISSUE #71: Add support for scikit-build build system

### PR DESCRIPTION
**Pull request by ayeankit**
## Pull Request description

Fixes #71 

1. Activate option in TUI to maturin
2. Adding documentation about scikit
3. Adding scikit as a build-system:
4. Creating a scikit-pyproject.toml
5. Editing post_gen_project.py
6. Creating a smoke test (build-system.sh)
7. Editing cookicutter.json
8. Added scikit in Readme.md

## How to test these changes

* ```...```

## Pull Request checklists

This PR is a:
- [ ] bug-fix
- [ ] new feature
- [ ] maintenance

About this PR:
- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [ ] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:
- [ ] I have reviewed the changes and it contains no misspelling.
- [ ] The code is well commented, especially in the parts that contain more complexity.
- [ ] New and old tests passed locally.

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

### CODE REVIEW

**Comment by xmnlab** [comment]
@ayeankit I just merged 2 PRs,  could you rebase your branch on top of upstream/main pls?

**Comment by xmnlab** [comment]
could you check if you are using the instructions from here?
https://scikit-build.readthedocs.io/en/latest/

File: README.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ for a Python package.
     [PDM](https://pdm.fming.dev/) or
     [Hatch](https://hatch.pypa.io)
     [Maturin](https://pypi.org/project/maturin/0.8.2/)
+    [scikit-hep](https://scikit-hep.org)
```
Comment (xmnlab) [inline README.md:37]: shouldn't it be scikit-build?

```suggestion
    [scikit-build](https://scikit-build.readthedocs.io/en/latest/)
```

File: src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md [hunk 2]
```diff
@@ -56,6 +56,9 @@ In addition, you should know that to build our package we use
...
+{%- elif cookiecutter.build_system == "maturin" -%}
+In addition, you should know that to build our package we use
+[scikit-hep](https://scikit-hep.org):It's a Python packaging tool and build system an improved build system generator for CPython C extensions. It provides better support for additional compilers, build systems, cross compilation, and locating dependencies and their associated build requirements.This tool improves package management in the scientific Python ecosystem, enabling cross-platform builds with CMake, and seamless integration with C/C++ libraries for research software engineers.
```
Comment (xmnlab) [inline src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md:61]: change scikit-hep to scikit-build text and refs, please

File: tests/smoke/build-system.sh [hunk 3]
```diff
@@ -9,3 +9,4 @@ SMOKE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 . ${SMOKE_DIR}/base.sh "build_system=pdm"
 . ${SMOKE_DIR}/base.sh "build_system=hatch"
 . ${SMOKE_DIR}/base.sh "build_system=maturin"
+. ${SMOKE_DIR}/base.sh "build_system=scikit"
```
Comment (xmnlab) [inline tests/smoke/build-system.sh:12]:
```suggestion
. ${SMOKE_DIR}/base.sh "build_system=scikit-build"
```

**Review by xmnlab — CHANGES_REQUESTED** [review]
@ayeankit thanks for working on that!
in general it looks good.
please change the references (names, links etc) from just `scikit` to `scikit-build`, also replace the references from `scikit-hep` by `scikit-build`

File: README.md (hunk 1 above)
Comment (ayeankit) [reply README.md:37]: actually I copied the link from `Scientific Python`, they have provided the they used skbuild for scikit.

Comment (xmnlab) [reply README.md]: yep skbuild is the name used to import scikit-build ... by the way, the scicookie from scientific python group is under `BSD 3-Clause License` if you are copying any code from there you need to ensure you are giving the credits (citation) properly

Comment (xmnlab) [reply README.md]: maybe you can find some information here: https://opensource.stackexchange.com/questions/10685/can-someone-re-license-my-bsd-3-licensed-project-under-the-mit-license-remove-m

Comment (xmnlab) [reply README.md]: but in general, I would say that would shouldn't copy any code from there .. just use it as reference

File: docs/guide.md [hunk 4]
```diff
@@ -325,7 +325,14 @@ packages. SciCookie support the following:
   compatibility with setuptools and Cargo make it an easy-to-use tool, offering
   developers a simple solution to combine the strengths of Python and Rust within
   a unified project.
-
+- [**scikit**](https://scikit-build.readthedocs.io/en/latest/):It's build system designed for Python
```
Comment (xmnlab) [inline docs/guide.md:328]:
```suggestion
- [**scikit-build**](https://scikit-build.readthedocs.io/en/latest/):It's build system designed for Python
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @ayeankit! This PR confuses `scikit-build` with `scikit-build-core`. I can see that you use `scikit-build-core` in `pyproject.toml`, but you are calling it `scikit-build` everywhere in this PR. [`scikit-build`](https://scikit-build.readthedocs.io/en/latest/) and [`scikit-build-core`](https://scikit-build-core.readthedocs.io/en/latest/) are technically different build backends.

**Comment by ayeankit** [comment]
> Thanks, @ayeankit! This PR confuses `scikit-build` with `scikit-build-core`. I can see that you use `scikit-build-core` in `pyproject.toml`, but you are calling it `scikit-build` everywhere in this PR. [`scikit-build`](https://scikit-build.readthedocs.io/en/latest/) and [`scikit-build-core`](https://scikit-build-core.readthedocs.io/en/latest/) are technically different build backends.

So should I change `scikit-build` to `scikit-build-core`, as scikit-build uses setup.py and scikit-build-core uses main.cpp

**Comment by Saransh-cpp** [comment]
We should ideally have both of them in `scicookie`. Given that you have already done most of the stuff keeping `scikit-build-core` in mind as the backend, you can switch all the text references to `scikit-build-core`.

**Comment by ayeankit** [comment]
@xmnlab Can you please review it once again?

File: docs/guide.md [hunk 5]
```diff
@@ -325,7 +325,14 @@ packages. SciCookie support the following:
...
   a unified project.
-
+- [**scikit-build-core**](https://scikit-build-core.readthedocs.io/en/latest/):It's build system designed for Python
+  packaging tool, serving as an enhanced build system generator for CPython
+  C extensions greatly improves package management within the scientific Python
+  ecosystem. It offers superior support for diverse compilers, build systems,
+  cross-compilation, and efficient dependency locating with their associated
+  build requirements. With its capabilities, it facilitates cross-platform
+  builds using CMake and effortless integration with C/C++ libraries, making
+  it a valuable asset for research software engineers.
```
Comment (Saransh-cpp) [inline docs/guide.md:335]:
```suggestion

- [**scikit-build-core**](https://scikit-build-core.readthedocs.io/en/latest/): It's build system designed for Python
  packaging tool, serving as an enhanced build system generator for CPython
  C extensions greatly improves package management within the scientific Python
  ecosystem. It offers superior support for diverse compilers, build systems,
  cross-compilation, and efficient dependency locating with their associated
  build requirements. With its capabilities, it facilitates cross-platform
  builds using CMake and effortless integration with C/C++ libraries, making
  it a valuable asset for research software engineers. 
  
```

File: src/scicookie/cookiecutter.json [hunk 6]
```diff
@@ -50,7 +50,8 @@
     "setuptools",
     "pdm",
     "hatch",
-    "maturin"
+    "maturin",
+    "scikit-build"
```
Comment (Saransh-cpp) [inline src/scicookie/cookiecutter.json:54]:
```suggestion
    "scikit-build-core"
```

File: src/scicookie/hooks/post_gen_project.py [hunk 7]
```diff
@@ -65,6 +65,8 @@
 BUILD_SYSTEM = "hatch"
 {% elif cookiecutter.build_system == "maturin" -%}
 BUILD_SYSTEM = "maturin"
+{% elif cookiecutter.build_system == "scikit-build" -%}
+BUILD_SYSTEM = "scikit-build"
```
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:69]:
```suggestion
BUILD_SYSTEM = "scikit-build-core"
```

File: src/scicookie/hooks/post_gen_project.py [hunk 8]
```diff
@@ -199,7 +201,20 @@ def clean_up_build_system():
         shutil.move(
             build_system_dir / "Cargo.toml",
             PROJECT_DIRECTORY / 'Cargo.toml'
-        )         
+        ) 
+    elif BUILD_SYSTEM == "scikit-build":
```
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:205]:
```suggestion
    elif BUILD_SYSTEM == "scikit-build-core":
```

File: src/scicookie/profiles/base.yaml [hunk 9]
```diff
@@ -88,6 +88,7 @@ build_system:
     - pdm
     - hatch
     - maturin
+    - scikit-build
```
Comment (Saransh-cpp) [inline src/scicookie/profiles/base.yaml:91]:
```suggestion
    - scikit-build-core
```

File: src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml [hunk 10]
```diff
@@ -20,6 +20,9 @@ dependencies:
 {%- elif cookiecutter.build_system == "maturin" %}
   - maturin
   - rust
+{%- elif cookiecutter.build_system == "scikit-build" %}
+  - scikit-build
```
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/conda/dev.yaml:24]:
```suggestion
  - scikit-build-core
```

File: src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md [hunk 11]
```diff
@@ -56,6 +56,9 @@ In addition, you should know that to build our package we use
 {%- elif cookiecutter.build_system == "maturin" -%}
 In addition, you should know that to build our package we use
 [Maturin](https://pypi.org/project/maturin/0.8.2/):It's a Python packaging tool and build system for creating Python bindings from Rust projects. It enables seamless integration of Rust code into Python applications, offering efficient builds, cross-platform support, and compatibility with different Python versions. Maturin automates the process of generating Python modules that directly call Rust functions, leveraging Rust's performance and low-level capabilities in Python. With its easy-to-use interface and integration with setuptools and Cargo, Maturin provides a straightforward solution for developers seeking to combine the strengths of Python and Rust in a single project.
+{%- elif cookiecutter.build_system == "scikit-build" -%}
```
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md:59]:
```suggestion
{%- elif cookiecutter.build_system == "scikit-build-core" -%}
```

File: src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md [hunk 12]
```diff
@@ -56,6 +56,9 @@ In addition, you should know that to build our package we use
...
+In addition, you should know that to build our package we use
+[scikit-build-core](https://scikit-build-core.readthedocs.io/en/latest/):It's a Python packaging tool and build system an improved build system generator for CPython C extensions. It provides better support for additional compilers, build systems, cross compilation, and locating dependencies and their associated build requirements.This tool improves package management in the scientific Python ecosystem, enabling cross-platform builds with CMake, and seamless integration with C/C++ libraries for research software engineers.
```
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/docs/contributing.md:61]:
```suggestion
[scikit-build-core](https://scikit-build-core.readthedocs.io/en/latest/): It's a Python packaging tool and build system an improved build system generator for CPython C extensions. It provides better support for additional compilers, build systems, cross compilation, and locating dependencies and their associated build requirements.This tool improves package management in the scientific Python ecosystem, enabling cross-platform builds with CMake, and seamless integration with C/C++ libraries for research software engineers.
```

File: tests/smoke/build-system.sh [hunk 13]
```diff
@@ -9,3 +9,4 @@ SMOKE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 . ${SMOKE_DIR}/base.sh "build_system=pdm"
 . ${SMOKE_DIR}/base.sh "build_system=hatch"
 . ${SMOKE_DIR}/base.sh "build_system=maturin"
+. ${SMOKE_DIR}/base.sh "build_system=scikit-build"
```
Comment (Saransh-cpp) [inline tests/smoke/build-system.sh:12]:
```suggestion
. ${SMOKE_DIR}/base.sh "build_system=scikit-build-core"
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]

**Comment by ayeankit** [comment]
Hey @Saransh-cpp , I think there needs to be one more PR for scikit-build with using scikit-buildas backend for build-system.

**Comment by Saransh-cpp** [comment]
Yes!

File: src/scicookie/hooks/post_gen_project.py [hunk 14]
```diff
@@ -199,7 +201,20 @@ def clean_up_build_system():
         shutil.move(
             build_system_dir / "Cargo.toml",
             PROJECT_DIRECTORY / 'Cargo.toml'
-        )         
+        ) 
+    elif BUILD_SYSTEM == "scikit-build-core":
+        shutil.move(
+            build_system_dir / "scikit-pyproject.toml",
```
Comment (Saransh-cpp) [inline src/scicookie/hooks/post_gen_project.py:207]:
```suggestion
            build_system_dir / "scikit-build-core-pyproject.toml",
```

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/scikit-pyproject.toml
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/scikit-pyproject.toml:1]: Rename `scikit-build-core-pyproject.toml`

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/skcdemo.cpp [hunk 15]
```diff
@@ -0,0 +1,9 @@
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
+
+PYBIND11_MODULE(skcdemo, m) {
+    m.def("hello", [](){
+        py::print("Hello, scikit-build!");
```
Comment (Saransh-cpp) [inline src/scicookie/{{cookiecutter.project_slug}}/build-system/skcdemo.cpp:7]:
```suggestion
        py::print("Hello, scikit-build-core!");
```

File: tests/smoke/base.sh [hunk 16]
```diff
@@ -52,7 +52,9 @@ elif command -v pdm &> /dev/null; then
 elif command -v hatch &> /dev/null; then
   COMMAND_PREFIX="hatch run"
 elif command -v maturin &> /dev/null; then
-  pip install .  
+  pip install . 
+elif command -v scikit &> /dev/null; then
```
Comment (Saransh-cpp) [inline tests/smoke/base.sh:56]:
```suggestion
elif command -v scikit-build-core &> /dev/null; then
```

File: tests/smoke/base.sh [hunk 17]
```diff
@@ -52,7 +52,9 @@ elif command -v pdm &> /dev/null; then
 elif command -v hatch &> /dev/null; then
   COMMAND_PREFIX="hatch run"
 elif command -v maturin &> /dev/null; then
-  pip install .  
+  pip install .
```
Comment (Saransh-cpp) [inline tests/smoke/base.sh:55]:
```suggestion
  pip install -e .
```

File: tests/smoke/base.sh [hunk 18]
```diff
@@ -52,7 +52,9 @@ elif command -v pdm &> /dev/null; then
 elif command -v hatch &> /dev/null; then
   COMMAND_PREFIX="hatch run"
 elif command -v maturin &> /dev/null; then
-  pip install .  
+  pip install . 
+elif command -v scikit &> /dev/null; then
+  pip install .
```
Comment (Saransh-cpp) [inline tests/smoke/base.sh:57]:
```suggestion
  pip install -e .
```

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
It would make sense to explicitly use `scikit-build-core` everywhere, given that you plan to add `scikit-build` as a backend too.

File: src/scicookie/{{cookiecutter.project_slug}}/build-system/scikit-pyproject.toml
Comment (ayeankit) [reply src/scicookie/{{cookiecutter.project_slug}}/build-system/scikit-pyproject.toml:1]: sure!!

File: tests/smoke/base.sh [hunk 19]
```diff
@@ -52,7 +52,9 @@ elif command -v pdm &> /dev/null; then
 elif command -v hatch &> /dev/null; then
   COMMAND_PREFIX="hatch run"
 elif command -v maturin &> /dev/null; then
-  pip install .  
+  pip install -e .
+elif command -v scikit-build-core &> /dev/null; then
```
Comment (xmnlab) [inline tests/smoke/base.sh:56]: are you sure that this command really exist?

I tried the latest version, and it doesn't have this command. 
so I would recommend something like:

```suggestion
elif [ "$(pip list|grep -c scikit_build_core)" -ne "0" ]; then
```

PS: please ensure always to check if the command exists or not before adding it here.

**Review by xmnlab — CHANGES_REQUESTED** [review]

File: tests/smoke/base.sh (hunk 19 above)
Comment (ayeankit) [reply tests/smoke/base.sh:56]: But how it worked for scikit-build?

Comment (xmnlab) [reply tests/smoke/base.sh]: not sure .. but you should always test it locally to be sure that the command works properly

**Review by xmnlab — APPROVED** [review]
thanks for working on that @ayeankit !
@Saransh-cpp thank you so much for the review!

**Pull request merged**